data/blog/2025/2025-03-18-mezzio101-what-defines-a-middleware-architecture.md

```diff
@@ -0,0 +1,123 @@
+---
+
+Nowadays, applications are getting more and more complex.
+The classic way of executing sequential code, from **request to response**, still works for simple apps, but using a modern architecture has many benefits.
+Enter the **middleware**, an intermediary code that processes the request in various ways before it even reaches your custom code.
+
+<!--- EXTENDED -->
+
+In this article we will list the advantages of using middlewares in your applications.
+From maintainability, to future-proof design, the benefits can't be ignored.
+
+### What does middleware do?
+
+The versatility of middleware is virtually endless.
+
+- Do you need it to pre-process your request by validating the user session or the form input?
+- Do you need to cache or translate the request?
+- Do you need to contact a third party site to enhance the request?
+
+Your custom middleware can handle **any specific logic** your application may require.
+Some useful things to implement are error handling, CORS, routing, translating.
+
+Ultimately, you can split your application into multiple middlewares that process the request before it reaches your main code.
+An added benefit is that the middlewares can be **reused more easily** in other applications.
+Even better, the middlewares **don't need to be written in the same programming language** or coding architecture.
+
+The alternative to the middleware architecture is the MVC architecture which has grown considerably since its inception, is feature rich, modular and reusable.
+Instead of using flexible layers like a middleware, MVC revolves around controllers and actions.
+Unfortunately, MVC's complexity has brought about its downfall, though it is still being used.
+The important bit and the central difference to middleware is that in a MVC, a request targets a single controller.
+Nowadays, the aim is more toward lightweight, modular and standardized frameworks for producing microframeworks and API based solutions which are more suited to a middleware approach.
+
+### How do you set up a middleware pipeline?
+
+The **middleware pipeline** defines the sequence in which the middlewares are executed.
+Setting up the pipeline needs to be done with care to avoid unwanted behavior.
+In some cases, the execution of one middleware can affect another middleware because of the way the request is passed along.
+The main things to consider when setting up the pipeline are:
+
+- Security
+- Performance
+- Functionality
+
+For example, it makes sense to add an error handler early in the pipeline, even as the first entry.
+If your application requires the user to be logged in, it doesn't make sense to process the request before the authentication middleware gives the go-ahead.
+Also, it doesn't make sense to have an authorization middleware placed before an authentication middleware in the pipeline.
+
+Based on the future needs of your application, the **pipeline can be easily updated** to include additional middleware.
```

Comment:
Explain this in more detail: why is it easier? Maybe provide an example, and potentially show how the same functionality in an MVC either requires deep ties into the MVC framework, or functionality that cannot be re-used.

Reply:
i expanded this section with an example and more on the MVC alternative